presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
     @Override
     public boolean pushdownFilterAndProjection(PageSourceOptions options)
     {
+        ImmutableList.Builder<FilterFunction> remainingFilterFunctions = ImmutableList.builder();
```

Comment:
I think it would be clear if this were factored out into several methods:

1) `pushdownFilterAndProjectionToPartitionColumns` (or similar name)
2) A method to extract out `PageSourceOptions` that will be sent to the delegate

And then we call method 1, get the result from method 2, and then send it to the delegate.

Reply:
So you would need to be able to get both the remaining filter functions, and also whether or not it failed.  I wonder if this could be simplified into empty list, so whenever remaining filter functions returns empty list (as it would if pushdown failed), then we trivially don't attempt to delegate.

Reply:
@tdcmeehan Here is what this might look like:

```
    @Override
    public boolean pushdownFilterAndProjection(PageSourceOptions options)
    {
        List<FilterFunction> remainingFilterFunctions = pushdownFilterOnPrefilledColumns(options);

        if (filterOnPrefilledValuesFailed) {
            return true;
        }

        this.outputChannels = Arrays.copyOf(options.getOutputChannels(), options.getOutputChannels().length);

        filterAndProjectPushedDown = delegate.pushdownFilterAndProjection(createPageSourceOptionsForDelegate(options, remainingFilterFunctions));
        return filterAndProjectPushedDown;
    }

    private List<FilterFunction> pushdownFilterOnPrefilledColumns(PageSourceOptions options)
    {
        ImmutableList.Builder<FilterFunction> remainingFilterFunctions = ImmutableList.builder();
        for (FilterFunction filterFunction : options.getFilterFunctions()) {
            Set<ColumnMappingKind> inputKinds = getInputKinds(filterFunction);
            if (!inputKinds.contains(PREFILLED)) {
                remainingFilterFunctions.add(filterFunction);
                continue;
            }

            if (inputKinds.size() > 1) {
                // TODO Pass necessary pre-filled values to the delegate so it can evaluate these filters
                throw new UnsupportedOperationException("Filters on a mix of partition and non-partition columns are not supported");
            }

            if (!filterFunction.isDeterministic()) {
                throw new UnsupportedOperationException("Non-deterministic filters on partition columns are not supported");
            }

            if (!evaluateFilterFunctionOnPrefilledValues(filterFunction)) {
                filterOnPrefilledValuesFailed = true;
                close();
                break;
            }
        }
        return remainingFilterFunctions.build();
    }
```